Data/Text/ICU/Spoof.hsc

```diff
@@ -0,0 +1,273 @@
+openFromSerialized :: ByteString -> IO MSpoof
+openFromSerialized b =
+  case toForeignPtr b of
+    (ptr, off, len) -> withForeignPtr ptr $ \p ->
+      wrapWithSerialized ptr =<< handleError (uspoof_openFromSerialized (p `plusPtr` off) (fromIntegral len) nullPtr)
+
+-- | Open a spoof checker given the contents of the "confusables.txt" and "confusablesWholeScript.txt"
+-- files as described in Unicode UAX #39.
+openFromSource :: ByteString -> ByteString -> IO MSpoof
+openFromSource confusables confusablesWholeScript =
+  unsafeUseAsCStringLen confusables $ \(cptr, clen) ->
+    unsafeUseAsCStringLen confusablesWholeScript $ \(wptr, wlen) ->
+      wrap =<< handleError (uspoof_openFromSource cptr (fromIntegral clen) wptr (fromIntegral wlen) nullPtr nullPtr)
+
+-- | Get the checks performed by a spoof checker.
+getChecks :: MSpoof -> IO [SpoofCheck]
+getChecks s = do
+  withSpoof s $ \sptr ->
+    (fromBitMask . fromIntegral . (.&.) #{const USPOOF_ALL_CHECKS}) <$> handleError (uspoof_getChecks sptr)
+
+-- | Configure the checks performed by a spoof checker.
+setChecks :: MSpoof -> [SpoofCheck] -> IO ()
+setChecks s c = do
+  withSpoof s $ \sptr ->
+    handleError $ uspoof_setChecks sptr . fromIntegral $ toBitMask c
+
+-- | Get the restriction level of a spoof checker.
+getRestrictionLevel :: MSpoof -> IO (Maybe RestrictionLevel)
+getRestrictionLevel s = do
+  withSpoof s $ \sptr ->
+    (highestValueInBitMask . fromIntegral) <$> uspoof_getRestrictionLevel sptr
+
+-- | Configure the restriction level of a spoof checker.
+setRestrictionLevel :: MSpoof -> RestrictionLevel -> IO ()
+setRestrictionLevel s l = do
+  withSpoof s $ \sptr ->
+    uspoof_setRestrictionLevel sptr . fromIntegral $ toBitMask l
+
+-- | Get the list of locale names allowed to be used with a spoof checker.
+-- (We don't use LocaleName since the root and default locales have no meaning here.)
+getAllowedLocales :: MSpoof -> IO [String]
+getAllowedLocales s = do
+  withSpoof s $ \sptr ->
+    splitLocales <$> (peekCString =<< (handleError (uspoof_getAllowedLocales sptr)))
+    where splitLocales = fmap (unpack . strip) . splitOn "," . pack
+
+-- | Get the list of locale names allowed to be used with a spoof checker.
+-- (We don't use LocaleName since the root and default locales have no meaning here.)
+setAllowedLocales :: MSpoof -> [String] -> IO ()
+setAllowedLocales s locs = do
+  withSpoof s $ \sptr ->
+    withCString (intercalate "," locs) $ \lptr ->
+      handleError (uspoof_setAllowedLocales sptr lptr)
+
+-- | Check if two strings could be confused with each other.
+areConfusable :: MSpoof -> Text -> Text -> IO SpoofCheckResult
+areConfusable s t1 t2 = do
+  withSpoof s $ \sptr ->
+    useAsPtr t1 $ \t1ptr t1len ->
+      useAsPtr t2 $ \t2ptr t2len ->
+        makeSpoofCheckResult <$>
+          handleError (uspoof_areConfusable sptr t1ptr (fromIntegral t1len) t2ptr (fromIntegral t2len))
+
+-- | Generate a re-usable "skeleton" to check if an identifier is confusable
```

Comment:
Another Haddock pro tip: double quotes get turned into an attempt to hyperlink. (Yes, Haddock predates all the other markup languages by years, so don't start :-) )

To correctly render double quotes, escape them with backslashes.